### ISSUE TITLE
Unable to build ds_ctcdecoder for PowerPC

### ISSUE THREAD

**Issue by NormanTUD** [mozilla/DeepSpeech#3379]
- **OS Platform and Distribution**: Ubuntu 18.04.3 LTS 4.14.0-115.19.1.el7a.ppc64le
- **TensorFlow installed from**: upstream
- **TensorFlow version**: TensorFlow/1.15.0-fosscuda-2019b-Python-3.7.4
- **Python version**: 3.7.4
- **GCC/Compiler version**: gcc (GCC) 8.3.0

I want to train DS on a PowerPC supercomputer node, so I need to get it to work there. There's a big last problem in my way and that is that I cannot get ds_ctcdecode to work properly. My first thought was just to compile it with ``make bindings``, but that did not work on PPC, because the swig binaries downloaded from mozilla are only for amd64. When I fixed this (i.e. installed swig 4.0.2. by their sources and disabled the download of it in the `definitions.mk`-file it did kind of work, but not fully. That is, it created a file called `_swigwrapper.so` which also has lots of symbols, but it's still missing some. For example, when trying to "import ds_ctcdecoder", I get:

> ImportError: /home/s3811141/.local/lib/python3.7/site-packages/ds_ctcdecoder/_swigwrapper.so: undefined symbol: PyClass_Type

What must I do so that the `_swigwrapper.so`-file gets created successfully, so that I can use `ds_ctcdecoder` on a PPC64LE node?

**Comment by lissyx** [comment]
This is not hardware we have access, and i dont have time to investigate. I have no idea why you have that missing symbol. Please also understand we are consumer of swig, so if the issue is here its really even more out of our hands.

**Comment by lissyx** [comment]
Btw googling for your missing symbol links to a lot of cases of py2/py3 mismatch. Please cross check you are using the proper versions at all time.

**Comment by lissyx** [comment]
Especially, using a virtualenv can help ensure the whole build process is consistent

**Comment by lissyx** [comment]
Another thing : our fork of swig is really needed only for nodejs bindings. For python i would advise to keep using your distro's one,and 18.04 should have 3.0.2 which should work well

**Comment by olafthiele** [comment]
I think I saw a post by someone on Discourse who made it work by some nest trick. Did you search there before posting here?

**Comment by olafthiele** [comment]
Sorry, just remembered that was for webrtcvad on PowerPC, but I remember that I could build everything else on PPC architecture just fine.

**Comment by NormanTUD** [comment]
Thanks for the replies so far.

@lissyx You were right that this was partly a python2/python3 problem. I changed python to python3 in the definitions.mk and this problem was solved, but there was another one. (See end of this post). But swig 3.0.x did not work because in the 3.0.x libraries the unordered_map does not work (it says at the beginning of the swig file that it does not and it produces tons of errors). Also, I'm using singularity, so the build process is all inside the container.

@olafthiele I did not know that, but I solved the webrtc problems on PPC myself and submitted it as a patch ( https://github.com/NormanTUD/webrtc )

My problem now is that, even though I was able to build the .so for and with python3 file, I still get:

ImportError: cannot import name '_swigwrapper' from 'ds_ctcdecoder' (/home/s3811141/.local/lib/python3.7/site-packages/ds_ctcdecoder/__init__.py)

**Comment by lissyx** [comment]
> ImportError: cannot import name '_swigwrapper' from 'ds_ctcdecoder' (/home/s3811141/.local/lib/python3.7/site-packages/ds_ctcdecoder/**init**.py)

Have you checked the lib swigwrapper is there ? 
Have you checked what the dynamic linker does with `LD_DEBUG=all` ? 

Anyway im sorry but its really on your side so far, nothing much i can do especially since il supposed to be on holidays. 

> You were right that this was partly a python2/python3 problem. I changed python to python3 in the definitions.mk and this problem was solved

That would make me question how you virtualenv is actually working. 



> I'm using singularity, so the build process is all inside the container

Container is unrelated to python virtualenv. I insists we do document to use it, and your reports would show you might have weird behavior just because of that.

**Comment by lissyx** [comment]
> I insists we do document to use it, and your reports would show you might have weird behavior just because of that

There's no good reason for you to have to change definitions.mk, for example. Virtualenv already takes care of making sure `python` matches what you have in the venv. Here you might end up with weird mixes.

**Comment by lissyx** [comment]
@NormanTUD Can you share any update ?

**Comment by NormanTUD** [comment]
@lissyx I've tried to do it with the original, unmodified build-files and as the manual suggested me to do it, but it did not work and I moved away from trying it on PPC because I spent way too much time trying to get it to work without any usable results.

**Comment by lissyx** [comment]
> but it did not work

How did it not work ?



> I spent way too much time trying to get it to work without any usable results

I'm really sorry about that, but it's not really hardware I can access to, so it's hard for me to assert what comes from ppc, what comes from your distro, what might be misdocumented.

**Comment by lissyx** [comment]
@NormanTUD I've had a look, no PPC on AWS EC2 nor on GCP.

**Comment by NanoNabla** [comment]
I know the problem is solved on the cluster of TUD and already reported it to @NormanTUD.

To put the solution on record for everybody else with the same problem on PowerPC.

```
# build your own SWIG
mv ds-swig ds-swig.orig
wget http://prdownloads.sourceforge.net/swig/swig-4.0.2.tar.gz
tar xfz swig-4.0.2.tar.gz
cd swig-4.0.2
./configure --prefix=`realpath ../ds-swig`
make
make install

cd ctcdecode
# in ../definitions.mk replace "PYTHON_PLATFORM_NAME := --plat-name manylinux1_x86_64" by "PYTHON_PLATFORM_NAME := --plat-name linux_ppc64le"
make bindings
pip install dist/ds_ctcdecoder-0.9.1-cp37-cp37m-linux_ppc64le.whl
```
For should be a better solution for `PYTHON_PLATFORM_NAME` than simply change it by hand, but it worked.

Ticket can be closed then.

**Comment by lissyx** [comment]
> # build your own SWIG
> mv ds-swig ds-swig.orig
> wget http://prdownloads.sourceforge.net/swig/swig-4.0.2.tar.gz
> tar xfz swig-4.0.2.tar.gz
> cd swig-4.0.2
> ./configure --prefix=`realpath ../ds-swig`
> make
> make install

Don't use that, it was documented (but it seems to have disappeared) to use https://github.com/lissyx/swig/tree/taskcluster (it includes fixed for nodejs)



> For should be a better solution for `PYTHON_PLATFORM_NAME` than simply change it by hand, but it worked.

So it was simply PYTHON_PLATFORM_NAME to change? Send a PR to make it overrideable from CLI, that would be welcome.

**Comment by NanoNabla** [comment]
> Don't use that, it was documented (but it seems to have disappeared) to use https://github.com/lissyx/swig/tree/taskcluster (it includes fixed for nodejs)
Thanks for this hint. We did not use nodejs, so there was no problem.

> So it was simply PYTHON_PLATFORM_NAME to change? Send
Changing `PYTHON_PLATFORM_NAME` was needed and building swig. 

The swig in `ds-swig` was not runable on PPC. Therefore I simply build it from source into a directory named the same `ds-swig`. Building ctcdecoder  was no problem with this swig building but the generated whl in (maybe corrupted?) manylinux1_x86_64 format. Changing `PYTHON_PLATFORM_NAME` fixed this problem.

I could supply a PR for this `PYTHON_PLATFORM_NAME` but I do not know how to deal for swig.

**Comment by lissyx** [comment]
> > 
> 
> The swig in `ds-swig` was not runable on PPC.

The prebuilt ones, it's expected, we don't have binaries for PPC



> Therefore I simply build it from source into a directory named the same `ds-swig`.

I just advise to use my tree for building



> but the generated whl in (maybe corrupted?) manylinux1_x86_64 format. Changing `PYTHON_PLATFORM_NAME` fixed this problem.

That's not really a bug, we force `manulinux1_x86_64` on any linux platform, I guess we could relax that and check the arch somehow, so we allow people to override `PYTHON_PLATFORM_NAME` as you need to.



> I could supply a PR for this `PYTHON_PLATFORM_NAME` but I do not know how to deal for swig.

Please do so, and for swig, rebuilding is fine, just build it from  https://github.com/lissyx/swig/tree/taskcluster. If you send a PR, it's a good thing to also ensure we document that.

**Comment by lissyx** [comment]
@NanoNabla Does that help you to write a PR ?

**Comment by NanoNabla** [comment]
I'm working on a proper PR. I did a commit on my fork https://github.com/tud-zih-tools/DeepSpeech/tree/ppc64le_integration but setting `SWIG_DIST_URL` seems not to work.
Simply running `SWIG_DIST_URL=https://[...] TARGET=ppc64le make bindings` results in echoing `There is no prebuilt SWIG available for ppc64le. Please produce one and set SWIG_DIST_URL`

Moreover, do I have to create a `config_setting` target in `BUILD`?

I'm not that familiar with your build system.

**Comment by lissyx** [comment]
> I'm not that familiar with your build system.

This is not ours per choice, this is a side-effect from using TensorFlow.



> Moreover, do I have to create a `config_setting` target in `BUILD`?

I dont think you need that to be able to build, and we have no mean to support a ppc64le builds, except if you can provide hw and babysit this, which I suspect you don't have bandwith as well :)



>  Simply running `SWIG_DIST_URL=https://[...] TARGET=ppc64le make bindings` results in echoing `There is no prebuilt SWIG available for ppc64le. Please produce one and set SWIG_DIST_URL`

Well, this is something we put in place to help people in your case, but since we don't use it it's not surprising it might not work as expected.

It's handled in the `Makefile`: https://github.com/mozilla/DeepSpeech/blob/master/native_client/definitions.mk#L209-L218



> I did a commit on my fork https://github.com/tud-zih-tools/DeepSpeech/tree/ppc64le_integration but setting `SWIG_DIST_URL` seems not to work.

Since you change code, you should have
```
ifeq ($(TARGET),ppc64le)
SWIG_DIST_URL := url/to/your/build
else
```

@NanoNabla Maybe just change all `:=` to `?=` so `make` can  pick your `SWIG_DIST_URL` from CLI? And do the same for `PYTHON_PLATFORM_NAME` so you don't have to change `TARGET` (because it's going to introduce other issues, in your case the correct value is `TARGET=host`)

**Comment by lissyx** [comment]
@NanoNabla This diff should fix being able to override the `SWIG_DIST_URL`:
```
diff --git a/native_client/definitions.mk b/native_client/definitions.mk
index 69fe4539..b2f8a985 100644
--- a/native_client/definitions.mk
+++ b/native_client/definitions.mk
@@ -207,6 +207,7 @@ define copy_missing_libs
 endef
 
 SWIG_DIST_URL ?= 
+ifeq ($(SWIG_DIST_URL),)
 ifeq ($(findstring Linux,$(OS)),Linux)
 SWIG_DIST_URL := "https://community-tc.services.mozilla.com/api/index/v1/task/project.deepspeech.swig.linux.amd64.1a4c14945012f1282c2eddc174fb7674d5295de8.0/artifacts/public/ds-swig.tar.gz"
 else ifeq ($(findstring Darwin,$(OS)),Darwin)
@@ -215,7 +216,8 @@ else ifeq ($(findstring _NT,$(OS)),_NT)
 SWIG_DIST_URL := "https://community-tc.services.mozilla.com/api/index/v1/task/project.deepspeech.swig.win.amd64.1a4c14945012f1282c2eddc174fb7674d5295de8.0/artifacts/public/ds-swig.tar.gz"
 else
 $(error There is no prebuilt SWIG available for your platform. Please produce one and set SWIG_DIST_URL.)
-endif
+endif # findstring()
+endif # ($(SWIG_DIST_URL),)
 
 # Should point to native_client/ subdir by default
 SWIG_ROOT ?= $(abspath $(shell dirname "$(lastword $(MAKEFILE_LIST))"))/ds-swig
```

I'll let you fix also the case of `PYTHON_PLATFORM_NAME` as suggested and test that on your side, and I'll welcome it as a PR.

**Comment by lissyx** [comment]
Fixed by #3505. @NanoNabla Don't hesitate to send new PR for the doc counter part to help people how to do this in the future :)
